### ISSUE TITLE
Phase 1: Vulkan Compute Foundation

### ISSUE THREAD

**Issue by ulises-jeremias** [vlang/vtl#58]
## Motivation

VTL currently has zero GPU compute capability. The entire training loop (forward, loss, backward, optimizers) runs on CPU. Conv2D is a 7-deep nested loop — the primary bottleneck. This phase establishes the **Vulkan Compute foundation** — the cross-vendor GPU backend that enables GPU-accelerated training on NVIDIA, AMD, Intel, ARM Mali/Adreno, Qualcomm, and Imagination GPUs with a single code path.

Vulkan Compute is chosen over CUDA as the primary target because:
- **Cross-vendor**: One implementation works on all GPU vendors (no proprietary SDK required)
- **Native ARM support**: Covers Android (ARM Mali/Adreno), iOS (via Metal interop), embedded
- **SPIR-V**: Unified GPU IR across all vendors — same kernels compile for all hardware
- **No vendor lock-in**: Unlike CUDA which only works on NVIDIA

### Reference Repositories

The V ecosystem has existing GPU/compute infrastructure used as **reference**:

- **`antono2/vulkan`** — Full raw Vulkan 1.0–1.4 bindings (~1.3 MB, weekly auto-regeneration from Khronos XML). Use as the **pattern** for VSL's own Vulkan bindings (`vsl/vulkan/`): C→V type mapping, struct layout, handle definitions, API function signatures. MIT licensed.
- **`antono2/v_vulkan_bindings`** — Python generator: Khronos XML → V code. Fork to generate VTL-specific Vulkan bindings subset. MIT licensed.
- **`antono2/vulkan_memory_allocator`** — Pool-based GPU memory allocator. Use as **architectural reference** for VTL's memory allocator. MIT licensed.
- **`vsl.vcl`** — Mature OpenCL wrapper. Use as **pattern** for compute abstraction and async buffer management. MIT licensed.

### Implementation Pattern: Self-Contained Wrappers

**Decision:** VTL/VSL maintains its own Vulkan bindings within `vsl/vulkan/` (VSL) and `vkl/` (VTL wrapper), following the same pattern as BLAS, LAPACK, and VCL. Use `antono2/vulkan`, `antono2/v_vulkan_bindings`, and `antono2/vulkan_memory_allocator` as reference implementations.

This keeps both VTL and VSL self-contained, avoids external dependencies on third-party modules, and gives full control over the Vulkan API surface.

### Files to create

1. **`vkl/`** — New V module: Vulkan compute bindings for VTL (wraps `vsl/vulkan/` or is standalone)
   - `vkl/module.v` — Module entry point
   - `vkl/device.v` — Device discovery, instance, physical device, logical device, queue
   - `vkl/buffer.v` — Buffer creation, memory binding
   - `vkl/memory.v` — Memory allocation, host<->device mapping, staging
   - `vkl/shader.v` — ShaderModule creation from SPIR-V
   - `vkl/pipeline.v` — Compute pipeline, pipeline layout
   - `vkl/descriptor.v` — Descriptor set layout, pool, allocation
   - `vkl/command.v` — Command buffer, submit, wait
2. **`storage/vulkan.v`** — `VulkanStorage[T]` wrapping Vulkan buffer, with `vcl()`-style transfer API
3. **`vulkan_tensor.v`** — `VulkanTensor[T]` mirror struct (shape, strides, memory format)
4. **`compute/gemm.v`** — `gemm_vulkan[T](a, b)` — matrix multiplication via Vulkan compute shader (16x16 tiled)
5. **`compute/elementwise.v`** — `relu_vulkan`, `sigmoid_vulkan`, `tanh_vulkan`, `gelu_vulkan` on Vulkan
6. **`compute/broadcast.v`** — `broadcast_add_vulkan`, `broadcast_mul_vulkan` on Vulkan
7. **`compute/gemv.v`** — `gemv_vulkan[T](a, b)` — matrix-vector multiplication
8. **`vk/kernels.v`** — Embedded SPIR-V/GLSL compute shader sources as V string constants

### Files to modify

- **`tensor_any_vulkan.v`** — Expand `AnyTensor[T]` to include `VulkanTensor[T]`
- **`tensor_any_notd_vulkan.v`** — Stub for when Vulkan is disabled
- **`nn/layers/linear.v`** — Add `$if defined(vulkan)` dispatch for `forward()`
- **`nn/layers/conv2d.v`** — Add Vulkan path (Phase 2, but wire up the dispatch in Phase 1)

### API Contracts (from storage to ops)

```v
// Storage transfer
pub fn (t &Tensor[T]) vulkan() !&VulkanTensor[T]           // CPU -> GPU
pub fn (t &VulkanTensor[T]) cpu() !&Tensor[T]               // GPU -> CPU
pub fn (t &VulkanTensor[T]) release() !                    // Free GPU memory

// Compute primitives
pub fn gemm_vulkan[T](a &Tensor[T], b &Tensor[T], ...)!&Tensor[T]
pub fn gemv_vulkan[T](a &Tensor[T], b &Tensor[T]) !&Tensor[T]
pub fn relu_vulkan[T](x &VulkanTensor[T]) !&VulkanTensor[T]
pub fn sigmoid_vulkan[T](x &VulkanTensor[T]) !&VulkanTensor[T]
pub fn tanh_vulkan[T](x &VulkanTensor[T]) !&VulkanTensor[T]
pub fn gelu_vulkan[T](x &VulkanTensor[T]) !&VulkanTensor[T]
```

### Kernel Sources to Embed (SPIR-V compatible GLSL)

1. **GEMM kernel** (16x16 tile, row-major, float32)
2. **GEMV kernel** (1D dispatch, row-major, float32)
3. **ReLU kernel** (element-wise, float32)
4. **Sigmoid kernel** (element-wise, float32)
5. **Tanh kernel** (element-wise, float32)
6. **GELU kernel** (element-wise, float32)
7. **Broadcast Add kernel** (vector + scalar or vector + vector)
8. **Broadcast Mul kernel**

### SPIR-V Compilation Strategy

Use V's existing `v shader` tool or call `glslangValidator` directly:
- Kernels stored as V string constants in `vk/kernels.v`
- At runtime (or build time), compile to SPIR-V via `glslangValidator -V -x -o out.spv`
- Load SPIR-V into Vulkan via `vkCreateShaderModule`

### Testing Plan

- Unit tests for `VulkanStorage[T]` (create, upload, download, release)
- Unit tests for `VulkanTensor[T]` (shape, strides, contiguity checks)
- Integration test: `gemm_vulkan` produces same output as `la.matmul` (within FP32 tolerance)
- Integration test: element-wise ops match CPU implementations
- Benchmark: GEMM on Vulkan achieves >= 80% of theoretical peak (roofline analysis)
- CI: Vulkan tests run on Linux (NVIDIA/AMD/Intel), skip on macOS/Windows if no Vulkan driver

### Dependencies

- `vkl` module (new — needs to be written first)
- `vsl/vulkan/` Vulkan bindings (Phase A of VSL, #237) or standalone VTL bindings
- Existing VSL VCL infrastructure for reference on async buffer management
- `glslangValidator` in PATH for SPIR-V compilation (can be installed via package manager or vendored)

## Checklist

- [ ] `vkl/` module: device discovery, instance, physical device, logical device, queue
- [ ] `vkl/` module: buffer creation, memory binding, host<->device transfer
- [ ] `vkl/` module: shader module creation from SPIR-V
- [ ] `vkl/` module: descriptor set layout, pipeline layout, compute pipeline
- [ ] `vkl/` module: command buffer, submit, wait
- [ ] `storage/vulkan.v`: `VulkanStorage[T]` with create/from_array/to_array/release
- [ ] `vulkan_tensor.v`: `VulkanTensor[T]` struct
- [ ] `vk/kernels.v`: all GLSL compute shader sources
- [ ] `compute/gemm.v`: `gemm_vulkan[T]`
- [ ] `compute/gemv.v`: `gemv_vulkan[T]`
- [ ] `compute/elementwise.v`: relu, sigmoid, tanh, gelu
- [ ] `compute/broadcast.v`: broadcast add/mul
- [ ] `tensor_any_vulkan.v`: AnyTensor includes VulkanTensor
- [ ] `tensor_any_notd_vulkan.v`: stub when Vulkan disabled
- [ ] `nn/layers/linear.v`: forward on Vulkan
- [ ] Tests: storage, tensor, gemm, elementwise, broadcast
- [ ] CI: Vulkan test pipeline

Related: VSL #237 (Phase A: Vulkan Compute Backend for VSL Matrix/Vector)
Parent: #57
Labels: enhancement, gpu, phase-1

**Comment by ulises-jeremias** [comment]
## Progress Update — GPU Compute Ops Complete

### What was built (VTL Phase 1)

**Storage layer:**
- ✅ `storage/vulkan_d_vulkan.v` — `VulkanStorage[T]` wrapping VSL `GpuBuffer`, with `vulkan()` / `cpu()` transfer and shared device management

**Tensor layer:**
- ✅ `tensor_vulkan_d_vulkan.v` — `VulkanTensor[T]` struct with shape/strides/memory/contiguity, `vulkan()` / `cpu()` transfer methods

**GPU Compute Operations (on `VulkanTensor[T]`):**

| Op | f32 (direct GPU dispatch) | f64 (VSL compute bridge) |
|----|---------------------------|--------------------------|
| `gemm(b)` | ✅ via `vsl.vulkan.gemm()` — zero CPU round-trip | ✅ via `vsl.compute.gemm_gpu()` (f64→f32→GPU→f32→f64) |
| `relu()` | ✅ via `vsl.vulkan.relu()` | ✅ via `vsl.compute.relu_matrix()` |
| `sigmoid()` | ✅ via `vsl.vulkan.sigmoid()` | ✅ via `vsl.compute.sigmoid_matrix()` |

### All 10 integration tests passing

```
GEMM f32 2x2, GEMM f64 2x2, GEMM f32 2x3×3x2,
ReLU f32/f64, Sigmoid f32/f64,
GEMM→ReLU chain, Dim mismatch error, Non-2D error
```

### Key architectural changes from original plan

1. **`vkl/` → `vulkan/` naming**: Renamed by user preference. VSL provides the `vsl.vulkan` module (device, buffers, pipelines, shaders), VTL builds on top.

2. **No separate `vkl/` module in VTL**: The Vulkan bindings live in VSL (`vsl/vulkan/`). VTL imports `vsl.vulkan` and `vsl.compute` directly.

3. **`AnyTensor[T]` interface NOT extended**: Adding `vulkan()` to `AnyTensor[T]` conflicts with VCL's `_d_vcl` / `_notd_vcl` conditional compilation suffixes (both `_d_vulkan` and `_d_vcl` need to coexist). **Solution**: `VulkanTensor[T]` is a concrete type — method calls go directly on it.

4. **Methods instead of standalone functions**: `vk_tensor.gemm(other)` instead of `gemm_vulkan(a, b)`. This fits the concrete-type approach.

5. **VSL compute layer for f64 bridging**: `VulkanTensor[f64]` ops use `vsl.compute` which handles f64↔f32 conversion transparently. f32 ops dispatch directly to GPU with zero CPU involvement.

### What remains

**Phase 1 remaining (low priority for MVP):**
- [ ] `compute/gemv.v` — matrix-vector multiply on Vulkan
- [ ] `compute/broadcast.v` — broadcast add/mul on Vulkan
- [ ] Full `v test` integration (requires `-enable-globals -d vulkan` flags)
- [ ] CI pipeline for Vulkan tests

**Phase 2 (blocked by interface resolution):**
- [ ] NN layer dispatch (linear → gemm on GPU, activations → relu/sigmoid on GPU)
  - Requires: resolving `AnyTensor[T]` interface conflict between Vulkan and VCL backends
- [ ] tanh/gelu kernels on GPU

### Related

- `vlang/vsl#237` ✅ Complete — VSL Vulkan Compute Backend (underlying infrastructure)
- All 10 integration tests verified manually with `v -enable-globals -d vulkan run`

**Comment by ulises-jeremias** [comment]
## Progress Update — GPU Compute Ops Complete

### What was built (VTL Phase 1)

**Storage layer:**
- ✅ `storage/vulkan_d_vulkan.v` — `VulkanStorage[T]` wrapping VSL `GpuBuffer`, with `vulkan()` / `cpu()` transfer and shared device management

**Tensor layer:**
- ✅ `tensor_vulkan_d_vulkan.v` — `VulkanTensor[T]` struct with shape/strides/memory/contiguity, `vulkan()` / `cpu()` transfer methods

**GPU Compute Operations (on `VulkanTensor[T]`):**

| Op | f32 (direct GPU dispatch) | f64 (VSL compute bridge) |
|----|---------------------------|--------------------------|
| `gemm(b)` | ✅ via `vsl.vulkan.gemm()` — zero CPU round-trip | ✅ via `vsl.compute.gemm_gpu()` (f64→f32→GPU→f32→f64) |
| `relu()` | ✅ via `vsl.vulkan.relu()` | ✅ via `vsl.compute.relu_matrix()` |
| `sigmoid()` | ✅ via `vsl.vulkan.sigmoid()` | ✅ via `vsl.compute.sigmoid_matrix()` |

### All 10 integration tests passing

```
GEMM f32 2x2, GEMM f64 2x2, GEMM f32 2x3×3x2,
ReLU f32/f64, Sigmoid f32/f64,
GEMM→ReLU chain, Dim mismatch error, Non-2D error
```

### Key architectural changes from original plan

1. **`vkl/` → `vulkan/` naming**: Renamed by user preference. VSL provides the `vsl.vulkan` module (device, buffers, pipelines, shaders), VTL builds on top.

2. **No separate `vkl/` module in VTL**: The Vulkan bindings live in VSL (`vsl/vulkan/`). VTL imports `vsl.vulkan` and `vsl.compute` directly.

3. **`AnyTensor[T]` interface NOT extended**: Adding `vulkan()` to `AnyTensor[T]` conflicts with VCL's `_d_vcl` / `_notd_vcl` conditional compilation suffixes (both `_d_vulkan` and `_d_vcl` need to coexist). **Solution**: `VulkanTensor[T]` is a concrete type — method calls go directly on it.

4. **Methods instead of standalone functions**: `vk_tensor.gemm(other)` instead of `gemm_vulkan(a, b)`. This fits the concrete-type approach.

5. **VSL compute layer for f64 bridging**: `VulkanTensor[f64]` ops use `vsl.compute` which handles f64↔f32 conversion transparently. f32 ops dispatch directly to GPU with zero CPU involvement.

### What remains

**Phase 1 remaining (low priority for MVP):**
- [ ] `compute/gemv.v` — matrix-vector multiply on Vulkan
- [ ] `compute/broadcast.v` — broadcast add/mul on Vulkan
- [ ] Full `v test` integration (requires `-enable-globals -d vulkan` flags)
- [ ] CI pipeline for Vulkan tests

**Phase 2 (blocked by interface resolution):**
- [ ] NN layer dispatch (linear → gemm on GPU, activations → relu/sigmoid on GPU)
  - Requires: resolving `AnyTensor[T]` interface conflict between Vulkan and VCL backends
- [ ] tanh/gelu kernels on GPU

### Related

- `vlang/vsl#237` ✅ Complete — VSL Vulkan Compute Backend (underlying infrastructure)
- All 10 integration tests verified manually with `v -enable-globals -d vulkan run`
